js/ui/codingManager.js

```diff
@@ -0,0 +1,638 @@
+                                                       rect.y,
+                                                       rect.width,
+                                                       rect.height);
+    },
+
+    _windowMinimized: function(shellwm, actor, session) {
+        // take the actor that we minimized and emit a minimized
+        // signal for it, though setting a flag internally so that
+        // we don't re-enter this signal handler.
+        if (this._processingWindowMinimized === actor) {
+            this._processingWindowMinimized = null;
+            return;
+        }
+
+        if (actor === session.actorApp && session.actorBuilder) {
+            this._processingWindowMinimized = session.actorBuilder;
+            session.actorBuilder.meta_window.minimize();
+        } else if (actor === session.actorBuilder && session.actorApp) {
+            this._processingWindowMinimized = session.actorApp;
+            session.actorApp.meta_window.minimize();
+        }
+    },
+
+    _windowUnminimized: function(shellwm, actor, session) {
+        // take the actor that we minimized and emit a minimized
+        // signal for it, though setting a flag internally so that
+        // we don't re-enter this signal handler.
+        if (this._processingWindowUnminimized === actor) {
+            this._processingWindowUnminimized = null;
+            return;
+        }
+
+        if (actor === session.actorApp && session.actorBuilder) {
+            this._processingWindowUnminimized = session.actorBuilder;
+            session.actorBuilder.meta_window.unminimize();
+        } else if (actor === session.actorBuilder && session.actorApp) {
+            this._processingWindowUnminimized = session.actorApp;
+            session.actorApp.meta_window.unminimize();
+        }
+    },
+
+    _windowAppRestacked: function(overview, stackIndices, session) {
+        let focusedWindow = global.display.get_focus_window();
+        if (!focusedWindow)
+            return;
+
+        // we get the signal for the same window switch twice
+        if (focusedWindow === session.previousFocusedWindow)
+            return;
+
+        // keep track of the previous focused window so
+        // that we can show the animation accordingly
+        let previousFocused = session.previousFocusedWindow;
+        session.previousFocusedWindow = focusedWindow;
+
+        // make sure we hide the button in any other case as on
+        // top of the App and Builder window
+        session.buttonApp.hide();
+        if (session.buttonBuilder)
+            session.buttonBuilder.hide();
+
+        if (focusedWindow === session.actorApp.meta_window) {
+            if (session.actorBuilder && session.actorBuilder.meta_window === previousFocused) {
+                // make sure we do not rotate when a rotation is running
+                if (this._rotateInActors.length || this._rotateOutActors.length) {
+                    session.buttonApp.show();
```

Comment:
I was commenting previously that this can go into a common code  path -- it's always called